### DISCUSSION TITLE
API implementation

### DISCUSSION THREAD

**Discussion by MaraScott** [Leantime/leantime#613] [Ideas]
Hi,

What would be the cost of integrating an API?

I m trying to find a solution to easily integrate emails in a help desk/todos workflow where tasks can become tickets and tickets become tasks as well as being able to centralize that data.
So far I m using freescout + leantime and I d like to connect these 2 awesome project via API but you Don t have one as far as I can read.

Best regards

David

**Comment by marcelfolaron** [comment]
We are slowly working to include API endpoints. A lot of the basic ones are already available however I haven't been able to build the API authentication yet.

**Comment by martydingo** [comment]
How has this been faring? I don't see any API documentation and it is a make or break decision, as I do want the ability to build middleware for integrating into other platforms.

**Comment by marcelfolaron** [comment]
We have put in place a json rpc api and I just finished writing the UI for
api key generation.

This will go out this week. I will create short docs right away and then
work on longer documentation over the next few weeks.

In summary you will be able to send get or post requests to /api/jsonrpc
with a method name from our services.

We’re following the jsonrpc standard v2
https://www.jsonrpc.org/specification

The code is in master already I just need to fix some unrelated bugs and
get the release out.

**Reply by gelven4sec** [reply]
Hi,

Thanks for implementing this !

I am having trouble working with jsonrpc, I don't think I'm using the right methods.

You've said you'll create some short docs, where can I find them ? I just need an example to get started.

**Reply by marcelfolaron** [reply]
Hi, 
thanks for your interest. 
I just pushed a quick summary on the usage of the API to our docs: https://docs.leantime.io/#/api/usage
We are still working on the docs for each individual service method. But this should get your started.

**Reply by gelven4sec** [reply]
Thanks a lot !

I'm just facing an odd issue, no matter what resquest I do, it response is : 
```
{success}
```

Even badly formatted json, GET or POST, anything.

**Reply by marcelfolaron** [reply]
Oh, you should not be getting that return in any case. What endpoint are you using? it should only be `{{domain}}/api/jsonrpc`

**Reply by gelven4sec** [reply]
That's the endpoint I'm using.

Any since I set up the `x-api-key` header, I only get that success message.

**Reply by marcelfolaron** [reply]
Oh, sorry can your try to run the request via /api/jsonRPC (note the capital RPC). I just looked through the code and it checks for that string and doesn't ignore case sensitivity. That would be a bug I'll have to fix here soon.

**Reply by gelven4sec** [reply]
I get a 404 on the `/api/jsonRPC` endpoint

**Reply by marcelfolaron** [reply]
Ugh, okay, sorry to give you the run around. I am fairly certain it's still related to that url check. Instead of using `/api/jsonRPC` please go back to using the lowercase variant but update core/class.application.php line 110
https://github.com/Leantime/leantime/blob/6d76840acccdc596e3a7cae93dc5de4503701a21/app/core/class.application.php#L101

to
```php
if(str_starts_with(frontController::getCurrentRoute(), "api.jsonrpc")){
```

**Reply by gelven4sec** [reply]
It works !
But only with a GET request.

In a POST request I get : 
```json
{
    "code": -32600,
    "message": "Invalid Request",
    "data": "Must include method"
}
```

The body contain this : 
```json
{
    "method": "leantime.rpc.tickets.getTicket",
    "jsonrpc": "2.0",
    "id": "1",
    "params": {"id":"9" }
}
```

**Reply by marcelfolaron** [reply]
That is odd. The error message get's sent when leantime cannot find the method parameter at all (empty). Can you make sure the parameter actually makes it to the system? How are you sending the post data right now?

**Reply by gelven4sec** [reply]
Alright so I got it working by sending parameters with form-data.
The documentation made me though I could just send the whole json in the body of the request.

One last thing before I leave you alone, where can find the complete list of supported methods ?
I guess that the domain of the method as to be one the services [here](https://docs.leantime.io/#/technical/README?id=leantimedomainservices)
But the related methods doesn't seem to exist. I tried "leantime.rpc.tickets.newTicket" which doesn't work.

**Reply by marcelfolaron** [reply]
We are still working on more detailed docs but the best way to get a list of available service methods is here: https://github.com/Leantime/docs/tree/master/technical/classes/leantime/domain/services

These docs will be updated but in the meantime feel free to ask about methods you need. You can also join our discord server to get answers faster. https://discord.gg/4zMzJtAq9z

**Reply by gelven4sec** [reply]
Exactly what I needed.
Thanks a lot for your help all along !

**Comment by mdmiko** [comment]
> Hi,
> 
> What would be the cost of integrating an API?
> 
> I m trying to find a solution to easily integrate emails in a help desk/todos workflow where tasks can become tickets and tickets become tasks as well as being able to centralize that data. So far I m using freescout + leantime and I d like to connect these 2 awesome project via API but you Don t have one as far as I can read.
> 
> Best regards
> 
> David

you were able to implement this integration? Can you share here?

**Comment by kalyan312** [comment]
I able to create task by 
curl http://localhost:8001/api/jsonrpc -XPOST -H "x-api-key: lt_na6mgGChzcI4hf2yUApjGR3waM0CBOxs_eRYSWO2cIvvLT3WR9p2Bco4ws01zAIoK"  -H "Content-Type: application/json" -d '{
    "method": "leantime.rpc.tickets.addTicket",
    "jsonrpc": "2.0",
    "values": {"id":11},
    "params": {"values":{"id":11,"headline":"Some task by API","type":"task","description":"<p>New Task<\/p>","projectId":3,"editorId":"","userId":1,"priority":"","sortIndex":null,"date":"2023-12-07T03:32:00-08:00","timelineDate":null,"timelineDateToFinish":null,"dateToFinish":"","timeToFinish":false,"status":3,"storypoints":0,"hourRemaining":0,"planHours":0,"sprint":0,"acceptanceCriteria":"","tags":"hello","url":null,"editFrom":"2023-12-08T00:00:00-08:00","timeFrom":"00:00:00","editTo":"2023-12-23T23:59:59-08:00","timeTo":false,"dependingTicketId":0,"parentHeadline":null,"milestoneid":0,"projectName":"Demo","clientName":"SoftBD ltd.","userFirstname":"Kalyan","userLastname":"Halder","editorFirstname":null,"editorLastname":null,"doneTickets":null,"allTickets":null,"percentDone":null,"milestoneHeadline":null,"milestoneColor":null,"editorProfileId":null,"bookedHours":null}}
    }' -v
